Require sqlite when building the wallet

Sjors opened the pull request:
Require that sqlite is available in order to compile the wallet. Removes instances of `USE_SQLITE` since it is no longer possible to not have sqlite available.

The `NO_SQLITE` option is dropped from depends.

This is another step towards dropping the legacy wallet, extracted from #31250.

jonatack commented:
Concept ACK, appears to be a standalone version of [`189eb8c` in #31250](https://github.com/bitcoin/bitcoin/pull/31250/commits/189eb8c02b5b81fe3239df34cd9ef48871e2e5de)

maflcko reviewed (commented):
There are a few leftovers. Genereally, when removing stuff, you can use git grep to find leftovers: `git grep WITH_SQLITE`. This also makes me wonder why CI passed. I would have expected for depends builds a warning:

```
CMake Warning:
  Manually-specified variables were not used by the project:
    WITH_SQLITE
```

maflcko reviewed (commented):
commit bafba0658c33e88664326f0b94a170f55bf7b5bd is wrong 🏃

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: commit bafba0658c33e88664326f0b94a170f55bf7b5bd is wrong 🏃
ocigj6g1m6WwMWXkk2SejazINqjyzq2RfdL1XdRULy9jkKlNNWH2A8yZQsONssjlwtN46BpdeR3jMlInu5c+Dg==
```

</details>

rkrux commented on test/functional/wallet_descriptor.py:
```diff
@@ -33,7 +33,6 @@ def set_test_params(self):
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
-        self.skip_if_no_sqlite()
         self.skip_if_no_py_sqlite3()
```
Should this call be removed as well? Right now, the `wallet_descriptor.py` test will be skipped if the python sqlite3 package is unavailable. I would assume that if the wallet is enabled (for which `skip_if_no_wallet()` 2 lines above is sufficient), I would want this test to run all the time instead of being skipped conditionally.

What do others think?

rkrux replied:
True.

My point was if the wallet is enabled, I would prefer to have this test run always for which this python package should be installed as well. But I now realise it is a developer's POV. 

Not removing this call would be acceptable if we assume not everyone who compile the wallet would want to run the functional tests, which I guess could be a big enough group.

Sjors commented on src/wallet/test/fuzz/CMakeLists.txt:
```diff
@@ -8,8 +8,7 @@ target_sources(fuzz
+    notifications.cpp
     spend.cpp
     wallet_bdb_parser.cpp
 )
```
The fuzz binary built on my machine, so I'm confused.

Sjors replied:
But anyway, I meant to drop only the duplicate. Will push fix.

Sjors commented:
Fixed https://github.com/bitcoin/bitcoin/pull/31961/commits/ac6cde731314d981391bc313c98d671c68211d33#r1975551868

hebasto commented:
Concept ACK.

furszy reviewed (commented):
utACK ac6cde731314d981391bc313c98d671c68211d33

brunoerg reviewed (approved):
code review ACK ac6cde731314d981391bc313c98d671c68211d33

hebasto commented:
Building depends with `NO_SQLITE=1` makes no sense now, doesn't it?

achow101 commented:
> Building depends with `NO_SQLITE=1` makes no sense now, doesn't it?

Makes sense if you don't want the wallet?

Sjors commented:
I could change depends here, but maybe it's easier to do that in the final PR that drops BDB itself?

hebasto commented:
> I could change depends here, but maybe it's easier to do that in the final PR that drops BDB itself?

Changes to depends shouldn't be big. Otherwise this statement from the PR description:
> ... it is no longer possible to not have sqlite available.

does not always hold (see a counter-example in https://github.com/bitcoin/bitcoin/pull/31961#issuecomment-2715187814).

fanquake commented:
> but maybe it's easier to do that in the final PR that drops BDB itself?

I don't see why, and it just leaves things broken/confusing in the interim. If we are going to move forward with this PR, the `NO_SQLITE` depends option should be removed at the same time.

Sjors commented:
I initially considered dropping the `NO_WALLET` option from depends. Most other depends packages refer to the library name rather than the feature, so this seemed better than dropping `NO_SQLITE`.

However it could break automations (and habits) of anyone who tries to build the node without wallet support. So I ended up dropping `NO_SQLITE` and keeping `NO_WALLET`.

Also rebased after #31161 just in case.

rkrux reviewed (commented):
Concept ACK 47e5ed3a5ba21049a9a8f3c4c44023e82e6ab88e

I've checked that there are no more instances present of the following make options, and `WITH_SQLITE` doesn't appear anymore in the output of `ccmake build`.

```
➜  bitcoin git:(2025/02/sqlite) ✗ git grep "USE_SQLITE"
➜  bitcoin git:(2025/02/sqlite) ✗ git grep "WITH_SQLITE"
➜  bitcoin git:(2025/02/sqlite) ✗ git grep "NO_SQLITE" 
```

Since the PR is rebased after the #31161 changes, I rm-rf'ed my build directory before configuring and building core and wallet again. I'm able to run bitcoind and bitcoin-cli successfully from the new `./build/bin` directory.

I also checked by explicitly passing `WITH_SQLITE=ON` again while configuring and noticed a warning like below.

```
➜  bitcoin git:(2025/02/sqlite) ✗ cmake -B build -DWITH_SQLITE=ON -DWITH_BDB=OFF -LH

-- Generating done (0.8s)
CMake Warning:
  Manually-specified variables were not used by the project:

    WITH_SQLITE
```

However, the `NO_WALLET=1` option (used like below) seemed to have no effect and I was able to use the wallet still. PLMK if I used this option incorrectly. It was only after I used `ENABLE_WALLET=OFF` that I was not able to use the wallet RPCs and instead received the `Method not found` error via the CLI.

```
cmake -B build -DNO_WALLET=1 -LH
```

Edit: I think this^ is not the right way because it is not a CMake build option, trying like below instead.
```
 make -C depends NO_WALLET=1
```

Edit 2: The `NO_WALLET` works fine as well when used while building depends like below:
```
make -C depends NO_WALLET=1 NO_QT=1 NO_QR=1 
cmake -B build --toolchain depends/x86_64-apple-darwin24.3.0/toolchain.cmake
cmake --build build -j 13
```

hebasto commented:
> However, the `NO_WALLET=1` option (used like below) seemed to have no effect and I was able to use the wallet still. PLMK if I used this option incorrectly. It was only after I used `ENABLE_WALLET=OFF` that I was not able to use the wallet RPCs.
> 
> ```
> cmake -B build -DNO_WALLET=1 -LH
> ```

`NO_WALLET=1` is dedicated to be used when building depends: `make -C depends NO_WALLET=1`.

hebasto reviewed (commented):
Approach ACK 47e5ed3a5ba21049a9a8f3c4c44023e82e6ab88e.

Here are two suggestions:
1. Simplify CMake code:
```diff
--- a/src/wallet/CMakeLists.txt
+++ b/src/wallet/CMakeLists.txt
@@ -28,6 +28,7 @@ add_library(bitcoin_wallet STATIC EXCLUDE_FROM_ALL
   rpc/wallet.cpp
   scriptpubkeyman.cpp
   spend.cpp
+  sqlite.cpp
   transaction.cpp
   wallet.cpp
   walletdb.cpp
@@ -37,17 +38,13 @@ target_link_libraries(bitcoin_wallet
   PRIVATE
     core_interface
     bitcoin_common
+    $<TARGET_NAME_IF_EXISTS:unofficial::sqlite3::sqlite3>
+    $<TARGET_NAME_IF_EXISTS:SQLite::SQLite3>
     univalue
     Boost::headers
     $<TARGET_NAME_IF_EXISTS:USDT::headers>
 )
 
-target_sources(bitcoin_wallet PRIVATE sqlite.cpp)
-target_link_libraries(bitcoin_wallet
-  PRIVATE
-    $<TARGET_NAME_IF_EXISTS:unofficial::sqlite3::sqlite3>
-    $<TARGET_NAME_IF_EXISTS:SQLite::SQLite3>
-)
 if(USE_BDB)
   target_sources(bitcoin_wallet PRIVATE bdb.cpp salvage.cpp)
   target_link_libraries(bitcoin_wallet PUBLIC BerkeleyDB::BerkeleyDB)
```

2. Adjust features in `vcpkg.json` as well:
```diff
--- a/vcpkg.json
+++ b/vcpkg.json
@@ -11,6 +11,7 @@
     "qt5",
     "tests",
     "wallet",
+    "berkeleydb",
     "zeromq"
   ],
   "features": {
@@ -28,12 +29,6 @@
         "libqrencode"
       ]
     },
-    "sqlite": {
-      "description": "Enable SQLite wallet support",
-      "dependencies": [
-        "sqlite3"
-      ]
-    },
     "tests": {
       "description": "Build test_bitcoin.exe executable",
       "dependencies": [
@@ -41,9 +36,8 @@
       ]
     },
     "wallet": {
-      "description": "Enable wallet",
+      "description": "Enable wallet (SQLite)",
       "dependencies": [
-        "berkeleydb",
         "sqlite3"
       ]
     },
```

Sjors commented:
Taken both suggestions.

hebasto commented:
> Taken both suggestions.
> 
> @hebasto native fuzz build is unhappy

```diff
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -171,7 +171,7 @@ jobs:
             generate-options: '-DBUILD_GUI=ON -DWITH_BDB=ON -DWITH_ZMQ=ON -DBUILD_BENCH=ON -DWERROR=ON'
             job-name: 'Win64 native, VS 2022'
           - job-type: fuzz
-            generate-options: '-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="sqlite" -DBUILD_GUI=OFF -DBUILD_FOR_FUZZING=ON -DWERROR=ON'
+            generate-options: '-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="wallet" -DBUILD_GUI=OFF -DBUILD_FOR_FUZZING=ON -DWERROR=ON'
             job-name: 'Win64 native fuzz, VS 2022'
 
     steps:
```

hebasto reviewed (approved):
ACK 564c54f3371b6faf50be2c7e7b6280357c90e62c.

Left a couple of nits and the following notes.

1. For some BSD systems, we document how to build BDB in depends. Now, it builds `sqlite` along with `bdb`. I don't believe it's critical, especially with BDB being removed soon.

2. re [comment](https://github.com/bitcoin/bitcoin/pull/31961#issuecomment-2717203847):
> I initially considered dropping the `NO_WALLET` option from depends. Most other depends packages refer to the library name rather than the feature, so this seemed better than dropping `NO_SQLITE`.

This alignes with the design as it was before `sqlite` was introduced in depends. However, in general, depends shouldn't be aware of the resulting software features.

Maybe @ryanofsky could provide more feedback on this?

Sjors committed:
build: require sqlite when building the wallet

Require that sqlite is available in order to compile the wallet. Removes
instances of USE_SQLITE since it is no longer possible to not have
sqlite available.

The NO_SQLITE option is dropped from depends.

Co-authored-by: Ava Chow <github@achow101.com>
Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

m3dwards commented:
Built with and without depends and all was working
 
 > I initially considered dropping the `NO_WALLET` option from depends. Most other depends packages refer to the library name rather than the feature, so this seemed better than dropping `NO_SQLITE`.
> 
> However it could break automations (and habits) of anyone who tries to build the node without wallet support. So I ended up dropping `NO_SQLITE` and keeping `NO_WALLET`.

As someone who doesn't have those habits my vote would have been for dropping `NO_WALLET`. I think of depends in terms of libraries, not bitcoin functionality.

 I assume this check in `src/leveldb/CMakeLists.txt` still makes sense rather than checking for `ENABLE_WALLET`?
 ```cmake
  check_library_exists(sqlite3 sqlite3_open "" HAVE_SQLITE3)
  if(HAVE_SQLITE3)
    leveldb_benchmark("benchmarks/db_bench_sqlite3.cc")
    target_link_libraries(db_bench_sqlite3 sqlite3)
  endif(HAVE_SQLITE3)
 ```

Sjors commented:
Pushed the two simplifications.

@hebasto the `NO_WALLET` option for depends has existed since at least v0.11: https://github.com/bitcoin/bitcoin/tree/v0.11.0/depends

m3dwards commented:
ACK 36b6f36ac4724cb2c9ed0e25314c3bbf55e4ebb8

hebasto reviewed (approved):
re-ACK 36b6f36ac4724cb2c9ed0e25314c3bbf55e4ebb8.

davidgumberg commented:
crACK https://github.com/bitcoin/bitcoin/commit/36b6f36ac4724cb2c9ed0e25314c3bbf55e4ebb8
